From fe6384633a14d487ecd39006a22a78eb2d9a6956 Mon Sep 17 00:00:00 2001
From: Saahil Tomar <saahtoma@qti.qualcomm.com>
Date: Wed, 3 Dec 2025 15:51:17 +0530
Subject: [PATCH] drivers: pci: qcom: Configure lane skew for dynamic lane
 switching

Set lane skew parameter to 0x20 to enable endpoint devices to dynamically switch between lanes at runtime. This lane switching capability has no functional impact on PCIe operation.

Change-Id: I3e53311f1aebe2676ec938d008521d09d07a2ed0
Signed-off-by: Saahil Tomar <saahtoma@qti.qualcomm.com>
---
 drivers/pci/controller/dwc/pcie-qcom.c | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/drivers/pci/controller/dwc/pcie-qcom.c b/drivers/pci/controller/dwc/pcie-qcom.c
index ae9670d59bb2..edf488e4e9d0 100644
--- a/drivers/pci/controller/dwc/pcie-qcom.c
+++ b/drivers/pci/controller/dwc/pcie-qcom.c
@@ -117,6 +117,7 @@
 #define LINK_CAPABLE_OFFSET(x)			((x) << 16)
 #define PCIE20_LANE_SKEW_OFF			0x714
 #define PCIE20_LANE_SKEW_OFF_MASK		0xFF000000
+#define PCIE20_LANE_SKEW_DYNAMIC_SWITCH_VAL	0x20
 #define PCIE20_MULTI_LANE_CONTROL_OFF		0x8C0
 #define PCIE20_LINK_CONTROL_LINK_STATUS_REG	0x80
 #define PCIE20_PARF_LTSSM_MASK			0x3F
@@ -1260,6 +1261,7 @@ static int qcom_pcie_post_init(struct qcom_pcie *pcie)
 	u16 offset = dw_pcie_find_capability(pci, PCI_CAP_ID_EXP);
 	u32 val;
 	int i;
+	int size = 4;
 
 	val = readl(pci->dbi_base + offset + PCI_EXP_DEVCTL);
 	val &= ~PCI_EXP_DEVCTL_PAYLOAD;
@@ -1321,6 +1323,11 @@ static int qcom_pcie_post_init(struct qcom_pcie *pcie)
 	writel(val, pci->dbi_base + offset + PCI_EXP_LNKCAP);
 #endif
 
+	/* Set lane skew parameter to enable dynamic lane switching at runtime */
+	dw_pcie_read(pci->dbi_base + PCIE20_LANE_SKEW_OFF, size, &val);
+	val = (val & PCIE20_LANE_SKEW_OFF_MASK) | PCIE20_LANE_SKEW_DYNAMIC_SWITCH_VAL;
+	dw_pcie_write(pci->dbi_base + PCIE20_LANE_SKEW_OFF, size, val);
+
 	writel(PCI_EXP_DEVCTL2_COMP_TMOUT_DIS, pci->dbi_base + offset +
 			PCI_EXP_DEVCTL2);
 
-- 
2.34.1

